### DISCUSSION TITLE
Sending messages is hard :)

### DISCUSSION THREAD

**Discussion by lurendrejer** [zwave-js/node-red-contrib-zwave-js#110] [Ideas]
Well, just saying 'hi' - had a great time moving my SH-setup from my current controller to NR with zwave-js.
I got everything going where something had to react to z-wave commands (turning on conbee lights, etc).

When I came to the part where i had to change states of z-wave devices themselves - I hit a wall.
Thermostats, binary switches, etc. - I tried and tried (https://github-wiki-see.page/m/zwave-js/node-red-contrib-zwave-js/wiki/CC-API)

I'm super impressed by how effecient zwave-js and node-red in itself is, but having to create z-wave messages manually is just too much for me.

If you ever consider creating some sort of function node, where one can select commands (multi, switch, whatever) - I'll be all over this thing.

Great work, thank you!

**Comment by lurendrejer** [comment]
oh, and the association-interface has a bug.
If i try to associate a multichannel device, with another i get this:
For instance: 31.1 [group 6] with 1.1 [controller]

I typed 1 and 1 into the interface.
It works if i associate with 1.0 (non multichannel) - but then the status updates doesn't work. They have to reach endpoint 1.1.

![image](https://user-images.githubusercontent.com/30922656/135111775-b10966d6-fb23-4192-b279-59e471e2fa46.png)

**Reply by marcus-j-davies** [reply]
> oh, and the association-interface has a bug. If i try to associate a multichannel device, with another i get this: For instance: 31.1 [group 6] with 1.1 [controller]
> 
> I typed 1 and 1 into the interface. It works if i associate with 1.0 (non multichannel) - but then the status updates doesn't work. They have to reach endpoint 1.1.
> 
> ![image](https://user-images.githubusercontent.com/30922656/135111775-b10966d6-fb23-4192-b279-59e471e2fa46.png)


1.1 is not a valid endpoint - endpoints are integers

**Reply by lurendrejer** [reply]
I typed "1" into the node ID and "1" into the endpoint - and got the above response from the UI.

**Reply by marcus-j-davies** [reply]
@lurendrejer 

ok,

Node 1 - is the controller, and the controller will only have endpoint 0.
try Node: 1, Endpoint: 0

**Reply by marcus-j-davies** [reply]
Also, there's no need to do this.
as the Controller will receive all events (i.e, it will already receive Lifeline i.e everything)

**Reply by lurendrejer** [reply]
It will be seen as a single-channel 'message'.
The controller will not know which sub-device sent the command.

These are working associations made with homeseer:
![image](https://user-images.githubusercontent.com/30922656/135132806-897ffed0-b78e-4a04-9b62-fb3fa9493808.png)

I can see the associations in zwave-js, but not create new ones like them.

**Reply by lurendrejer** [reply]
My controller has five endpoints, 1, 1.1, 1.2, 1.3 and 1.4

**Reply by marcus-j-davies** [reply]
The endpoint you set in the text box - is the destination endpoint, not the source endpoint.
the source endpoint, is the drop down.

not sure if that explains the difference?

**Reply by lurendrejer** [reply]
That much I did get,  :)

But the multichannel messages still needs to go to the controllers multichannel endpoint (1.1)

**Reply by marcus-j-davies** [reply]
> My controller has five endpoints, 1, 1.1, 1.2, 1.3 and 1.4
Thats the first I have come across - Controllers only have 1 endpoint (0)

I am actually checking specification on that now 😆 .....

**Reply by marcus-j-davies** [reply]
the controller will know what endpoint it come from,

the events actually detail the `endpoint` the node sent from

**Reply by lurendrejer** [reply]
Ok.
If a multichannel message is sent to 1.0 (1) - the controller will just update the 'main device' - not the correct sub-device.

**Reply by marcus-j-davies** [reply]
the controller should only ever have the devices **lifeline** pointed at it.

**Reply by lurendrejer** [reply]
I don't know if this plays any role when using zwave-js, but any other controller will fail to update subdevices in their guis.

**Reply by lurendrejer** [reply]
> the controller should only ever have the devices **lifeline** pointed at it.

Most devices, i am sure: https://devices.zwave-js.io/?jumpTo=0x0234:0x0003:0x010a:2.0 - Not these :)
They have a status association, per subdevice. (Basic Report)

**Reply by lurendrejer** [reply]
But again, it might not really matter - since zwave-js doesn't have a gui that it needs to keep in sync.

**Reply by marcus-j-davies** [reply]
node ID 1 - is the controller (it only has 1 endpoint - 0)

if you send a message to node 2, EP 0 - that is the device's "root device" root device is manufacturer specific (its not always black and white)

Example, if you have a node say **Node 6** with 2 switches.
it will likely have 2 EP's

EP 1 - first switch
Ep2 - second switch

so this.
```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        cc: "Binary Switch",
        endpoint: 2,
        node: 6,
        method: "set",
        params: [false]
    }
}
return Message
```

will turn of the 2nd switch on that device

at the same time, some devices allow certain events to be sent to another device (Node ID + its endpoint)
this is the associations.


All devices will have at least 1 Association (known as lifeline) - this addresses all its evenst to the controller regardless

**Reply by marcus-j-davies** [reply]
adding further associations addressed to the controller - should not be needed - as its all covers by the lifeline.

**Reply by marcus-j-davies** [reply]
The incoming message you get - will detail the endpoint that the device used to send i.e the source EP

**Reply by lurendrejer** [reply]
> adding further associations addressed to the controller - should not be needed - as its all covers by the lifeline.

Ok, i'm pretty much with you here...
But all other smart-home controllers I have tried, Fibaro, Homey, Vera, Homeseer, Homeassistant, etc. - fails to update the correct, virtual nodes - if they receive the report-messages on the main (1) controller node.

I'm drawing my conclusions on what I have seen, not what z-wave standards state.

**Reply by marcus-j-davies** [reply]
Any event hitting the controller node will update the UI (and by extension, the zwave-js value cache).
providing the devices Lifeline is pointed at the controller (N1 : EP:0)

the UI will show the last know values for each node/CC.

you can get the value cache :)

```javascript
let Message = {
    payload: {
        mode: "DriverAPI",
        method: "getValueDB",
        params: [...<NodeIDs>] /* Optional */
    }
}
return Message

```

**Reply by marcus-j-davies** [reply]
> > the controller should only ever have the devices **lifeline** pointed at it.
> 
> Most devices, i am sure: https://devices.zwave-js.io/?jumpTo=0x0234:0x0003:0x010a:2.0 - Not these :) They have a status association, per subdevice. (Basic Report)

<s>yup - but they will still be addressed to the controller (1, 0)</s>

the lifeline association for each device will cover (unless the manufacture is doing really odd things)
all values for all endpoints.

**Reply by lurendrejer** [reply]
I tried associating the multilevel association group with endpoint 1 (27.21 -> 1) - it doesn't update the value in the gui.

The association made with homeseer, to 1.1 does.

![image](https://user-images.githubusercontent.com/30922656/135138021-2745aa70-512d-4aca-82f6-f19dc19b8068.png)

**Reply by marcus-j-davies** [reply]
ok is this changing the device its self (or with a message)

**Reply by lurendrejer** [reply]
Physically, on the device - It should report back to the controller.
Associating with 1.1 from homeseer, plugging the controller back into zwave-js, makes it work.

**Reply by marcus-j-davies** [reply]
ok can you show me the life line association for that device?

**Reply by lurendrejer** [reply]
ofc
![image](https://user-images.githubusercontent.com/30922656/135139119-46626ae7-a778-49b9-8170-c45d879204aa.png)

**Reply by lurendrejer** [reply]
and the working, multichannel association: (Made with homeseer)
![image](https://user-images.githubusercontent.com/30922656/135139218-d78a21c9-da9f-4be9-bbe2-a3c6e9d922f6.png)

**Reply by marcus-j-davies** [reply]
Ok,

and when you tried that in the UI you had the indexOf error right?

**Reply by lurendrejer** [reply]
yup :)

**Reply by marcus-j-davies** [reply]
Ok,

Do you mind capturing this for me?
I am in contact with the zwave-js dev.

Remove the association you added for 1:1.

turn on logging (Debug) - attempt to re-add it (as above)
it will capture everything in a log file (no sensitive data is included)

if you can do this, Your a leg-end

**Reply by marcus-j-davies** [reply]
Attach the whole log file here.

**Reply by lurendrejer** [reply]
Does it have to be the same association or can i do it with any other multichannel group?

**Reply by marcus-j-davies** [reply]
if you can replicate the error - thats good enough 👍

**Reply by lurendrejer** [reply]
And how do i turn on logging and where does the log go? (ubuntu 21.04)

**Reply by marcus-j-davies** [reply]
Some devices, have special treatment, and it maybe this is one of them.
the logs will tell us - special treatment involves making a small change to the config for this device.

**Reply by marcus-j-davies** [reply]
On the controller node settings (**not** Pin 2 logging - that sends logging out to your flow)

log will be `~/.node-red/zwave-js.log`

**Reply by lurendrejer** [reply]
after enabling debug logging, the association interface is either dead or dead-slow... i'll wait and see.
![image](https://user-images.githubusercontent.com/30922656/135144671-5ddea588-7058-4201-839e-13269500dba0.png)

**Reply by marcus-j-davies** [reply]
you might have jumped the gun 😆 
refresh the UI.

**Reply by marcus-j-davies** [reply]
Once captured, it is best to turn off logging again - it will grow quickly 😁

**Reply by lurendrejer** [reply]
nopes... :)

2021-09-28T18:31:52.085Z NDERED   [ERROR] [INPUT] Node null does not exist.
2021-09-28T18:32:02.921Z NDERED « [CONTROLLER] Payload received.
2021-09-28T18:32:02.922Z NDERED « [API: ControllerAPI] [Method: getNodes]
2021-09-28T18:32:02.928Z NDERED » [NODE_LIST] Forwarding payload...
2021-09-28T18:32:10.528Z NDERED « [CONTROLLER] Payload received.
2021-09-28T18:32:10.528Z NDERED « [API: AssociationsAPI] [Method: getAllAssociationGroups]
                                  └─[params]
                                      0: null
2021-09-28T18:32:10.528Z NDERED   [ERROR] [INPUT] Node null does not exist.

**Reply by lurendrejer** [reply]
DOH! i didn't select a node!

**Reply by marcus-j-davies** [reply]
It happens to the best of us 😆

**Reply by lurendrejer** [reply]
[zwave-js1.log](https://github.com/zwave-js/node-red-contrib-zwave-js/files/7246564/zwave-js1.log)

**Reply by marcus-j-davies** [reply]
be sure to attach the whole log (as a file) - it will have raw message data in their for the device

**Reply by marcus-j-davies** [reply]
and you received the error?

**Reply by lurendrejer** [reply]
yes

**Reply by marcus-j-davies** [reply]
reading......

**Reply by lurendrejer** [reply]
Oh, and please never implement some sort of automatic association handling.
Domiticz does this - I startet it once and every association i had carefully removed from my network in homeseer, was suddently back.

**Reply by marcus-j-davies** [reply]
Right!

I have found the bug (thanks to the log)
it is indeed with the UI (I think) - A small change to the UI js file

fancy testing for me (if you have time)? - I can fix it tonight if so.

> Oh, and please never implement some sort of automatic association handling.

That one is for the zwave-js repo 😉

**Reply by marcus-j-davies** [reply]
In a nut shell.

this
`PL[2][0].endpoint = EI.val(parseInt(EI.val()));`

needs to be
`PL[2][0].endpoint = parseInt(EI.val())`

in the UI code.

**Reply by lurendrejer** [reply]
Oh, neat - thank you :)
I'm going to bed - I'm way behind on that.

I'll try it out after work tomorrow.

How do I update?

**Reply by marcus-j-davies** [reply]
No... Thank you.

I found the bug, and I am fixing it now - I will be publishing an update in the next hour (I will ping you when its up)
Once again - thanks for spotting that one.

https://github.com/zwave-js/node-red-contrib-zwave-js/issues/111

when its up - you will see it in the NR palette as an available update.

**Reply by crxporter** [reply]
Did you say bug 🐞

**Reply by marcus-j-davies** [reply]
Indeed!

when creating Multi Channel Associations I am taking the HTML element its self as the value - and not the value of the HTML Element.

🙈

**Reply by marcus-j-davies** [reply]
@lurendrejer.

6.1.1 has now been released, this fixes the multi channel association bug you found (again Thank you).

This should appear in the palette as an update.
Will be grateful if you can let me know if you can now add Multi Channel associations (I don't have any devices that make use of them)

**Comment by marcus-j-davies** [comment]
Hi @lurendrejer ,

lets take a look.

To turn on a Binary Switch.
https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/CC-API#examples
```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: 2,
        cc: "Binary Switch",
        method: "set",
        params: [true]
    }
}
return Message
```

**Reply by marcus-j-davies** [reply]
To adjust a Thermostat Set point

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: 2,
        cc: "Thermostat Setpoint",
        method: "set",
        enums: {0:"ThermostatSetpointType"},
        params: ["Heating", value, scale]
    }
}
return Message
```

**Comment by lurendrejer** [comment]
This is that I tried:
let Message = {
    payload: {
        mode: "CCAPI",
        node: 31,
        cc: "Binary Switch",
        endpoint: 2,
        method: "set",
        params: [true]
    }
}

does the return Message -part actually 'send' the command?

**Reply by marcus-j-davies** [reply]
> This is that I tried: let Message = { payload: { mode: "CCAPI", node: 31, cc: "Binary Switch", endpoint: 2, method: "set", params: [true] } }
> 
> does the return Message -part actually 'send' the command?

Does you device need an endpoint?

**Reply by lurendrejer** [reply]
Yes, it's a multichannel device.

**Reply by marcus-j-davies** [reply]
does it change when using the UI?

**Reply by lurendrejer** [reply]
Theres a ui? - sorry I completely missed that part.

**Reply by marcus-j-davies** [reply]
![image](https://user-images.githubusercontent.com/55892693/135115201-4b54b6d0-ab9a-44dd-86a2-f3bf433da446.png)

**Reply by lurendrejer** [reply]
Oooh I just thought that was a tool to look up the commands... doh!

**Reply by marcus-j-davies** [reply]
if it works here - let me know

**Comment by lurendrejer** [comment]
And how would i go about leaving 'node' empty, since i'm routing it to a specific z-wave node in my flow?

**Reply by marcus-j-davies** [reply]
just removing the node property from your message, will be enough.

**Comment by marcus-j-davies** [comment]
Just to re-cap 😆 

 - endpoints are integers (1,2,3,4)
  - let me know if the UI works (it might just be a funny set up with the device)
  
 I would also use the real wiki address.
 https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/getting-started

**Comment by lurendrejer** [comment]
How do i send the command from the UI?
Am i really that stupid?

**Reply by lurendrejer** [reply]
DOH! got it...

**Reply by marcus-j-davies** [reply]
The UI is sending the commands when you edit (click pencil)

![image](https://user-images.githubusercontent.com/55892693/135116580-db5f64c4-bdf5-4c57-8b6e-02c3db2e8be7.png)

**Reply by marcus-j-davies** [reply]
if it works - then no reason why it would not work with a message (which we can look at shortly)

**Reply by lurendrejer** [reply]
It does work :)
I'll get the message sorted tomorrow, I'm pretty sure of it.

Thank you for everything up until now, I'll be sure to update with my findings - if i find out the rest by myself, I might even be able to remember it -!

I was missing a crucial part - that I cound test-send messages and capture those.

Thank you, thank you, thank you!!

**Reply by marcus-j-davies** [reply]
double click  current value....

send me screen shot

**Reply by lurendrejer** [reply]
![image](https://user-images.githubusercontent.com/30922656/135117539-e61fb8dc-ee51-4601-a601-16688de6146a.png)

**Reply by marcus-j-davies** [reply]
based on that
```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: <your node id>, 
        endpoint: 3,
        cc: "Binary Switch",
        method: "set",
        params: [true]
    }
}
return Message
```

**Reply by marcus-j-davies** [reply]
the CC API does not update the UI.
so it will affect the device - but may not update the UI.

the CCAPI - is very low level,
the ValueAPI - whilst a little more advanced - updates the UI

**Comment by lurendrejer** [comment]
I'll be re-enabling my homeseer -> nodered -> conbee flow for the rest of the day, I think.

And look into the native zwave-js -> nodered -> conbee later.

After that, I'll start the zwave-js -> nodered -> native zigbee -project.

**Reply by lurendrejer** [reply]
Right now, I have too many layers of abstraction.

**Comment by lurendrejer** [comment]
And again, thank you so much for you time!

**Comment by marcus-j-davies** [comment]
OK,

Double clicking `current value` will show you the endpoint ID, and you can use that to correct the message.

Based on that.

in a function node -> Controller or zwave-device node

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: <your node id>, /* not needed for some modes on device-node /*
        endpoint: #,
        cc: "Binary Switch",
        method: "set",
        params: [true]
    }
}
return Message
```

if `endpoint` is 0 - you can omit `endpoint` from your message.

**Reply by lurendrejer** [reply]
I just omit the node: -part too? - if i'm sending it to a specific z-wave node-type-node.

**Reply by marcus-j-davies** [reply]
correct.
if set to **Specific Device** or **Multiple Device**

**Comment by marcus-j-davies** [comment]
Be sure to checkout the wiki in all its glory - it has examples for most things

**Reply by lurendrejer** [reply]
I will, and i thought i had.
Again, doh!

**Comment by marcus-j-davies** [comment]
Oh and welcome to z-wave for node red 😁 

@crxporter,
Is also part of the team

**Comment by lurendrejer** [comment]
It actually does seem to work... thank you.
But man it's a bit more tough than using the conbee-like interfaces in node-red.

![image](https://user-images.githubusercontent.com/30922656/135120282-979b10a4-b17a-403a-a7c6-aba1e92cda73.png)

**Reply by lurendrejer** [reply]
Since I have something working now, I think i'll create a subflow - that handles the basic, endpoint, etc. parts.
Like i did for the Central Scene commands.

![image](https://user-images.githubusercontent.com/30922656/135121224-d7de8a54-bdd3-42b6-96ec-3f43c99c20b5.png)

**Reply by marcus-j-davies** [reply]
Nice!

The device-node has the ability to use variables if inside a subflow.
and the one pot of gold for you will be the `event-filter` node.

you can send each different type of value update through a different output pin - i.e splitting the values out to different paths,

VERY handy for motion sensors with temp and lux sensing for an example.

do holla, if you want to know how.

**Reply by lurendrejer** [reply]
This is what i ended up with - it does the legwork :)

![image](https://user-images.githubusercontent.com/30922656/135124317-34c74fdc-6881-4863-a30c-67306bb3ee58.png)

**Reply by marcus-j-davies** [reply]
Looking good.

Here is the type of events to expect from the module.

https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/Payload-Messaging-Format#event-table
Scroll slightly up, to see how they are delivered.

**Comment by lurendrejer** [comment]
My simple test - triggered by a zigbee device, is working.
I'll look into what I can do with some advanced functionality (Basic, multilevel, Temperature etc.) tomorrow

![image](https://user-images.githubusercontent.com/30922656/135125546-499efadc-abae-4785-bdd8-cae75ca9a151.png)


once again, thank you!!

**Comment by marcus-j-davies** [comment]
Lastley,
feel free to join us on discord. 👍 .
we have a few deconz users here also

https://discord.gg/QD2Uu2cT

**Comment by crxporter** [comment]
If the original request here was "make commands easier" ... we are working on some ideas at #113

**Comment by lurendrejer** [comment]
Updating seems to fail:
2021-09-29T17:44:09.983Z Install : node-red-contrib-zwave-js 6.1.1

2021-09-29T17:44:09.738Z npm install --no-audit --no-update-notifier --no-fund --save --save-prefix=~ --production --engine-strict node-red-contrib-zwave-js@6.1.1
2021-09-29T17:44:18.978Z [err] npm ERR! code 1
2021-09-29T17:44:18.978Z [err] npm ERR! path /root/.node-red/node_modules/node-red-contrib-zwave-js
2021-09-29T17:44:18.979Z [err] npm
2021-09-29T17:44:18.980Z [err]  ERR! command failed
2021-09-29T17:44:18.980Z [err] npm ERR! command sh -c npm install --build-from-source --production @serialport/bindings@9.2.1 && npm install --build-from-source --production serialport@9.2.1
2021-09-29T17:44:18.980Z [err] npm ERR! npm WARN npm npm does not support Node.js v16.10.0
2021-09-29T17:44:18.980Z [err] npm ERR! npm WARN npm You should probably upgrade to a newer version of node as we
2021-09-29T17:44:18.980Z [err] npm ERR! npm WARN npm can't make any promises that npm will work with this version.
2021-09-29T17:44:18.980Z [err] npm ERR! npm WARN npm Supported releases of Node.js are the latest release of 4, 6, 7, 8, 9, 10.
2021-09-29T17:44:18.980Z [err] npm ERR! npm WARN npm You can find the latest version at https://nodejs.org/
2021-09-29T17:44:18.980Z [err] npm ERR! npm ERR! cb.apply is not a function
2021-09-29T17:44:18.980Z [err] npm ERR! 
2021-09-29T17:44:18.980Z [err] npm ERR! npm ERR! A complete log of this run can be found in:
2021-09-29T17:44:18.980Z [err] npm ERR! npm ERR!     /root/.npm/_logs/2021-09-29T17_44_18_543Z-debug.log
2021-09-29T17:44:18.987Z [err] 
2021-09-29T17:44:18.987Z [err] npm ERR! A complete log of this run can be found in:
2021-09-29T17:44:18.987Z [err] npm 
2021-09-29T17:44:18.987Z [err] ERR!     /root/.npm/_logs/2021-09-29T17_44_18_981Z-debug.log

**Reply by lurendrejer** [reply]
Just installed the 'unifi' addon - that seemed to go just fine

**Reply by lurendrejer** [reply]
npm@7.24.0 /usr/lib/node_modules/npm

**Reply by crxporter** [reply]
> Just installed the 'unifi' addon - that seemed to go just fine

The `node-red-contrib-unifi-os` ? You really should come by our discord... same discord is where we are discussing development of this unifi contrib...

I would love some more input from unifi users!

A fresh link: https://discord.gg/uvYac5u

**Reply by marcus-j-davies** [reply]
ok - thats a permissions thing (I think)

login to SSH as root, go to `/root/.node-red`
and do `sudo npm install --production node-red-contrib-zwave-js@latest`

**Reply by marcus-j-davies** [reply]
providing your **.node-red** folder is indeed in root?

**Reply by marcus-j-davies** [reply]
@lurendrejer 
have you done any `nodejs` or `npm` updates?
I dont know if this is due to a recent nodejs update or its just a permissions thing.

2021-09-29T17:44:18.980Z [err] npm ERR! npm **WARN** npm npm does not support Node.js v16.10.0

npm - isn't alway clear  as to the reason

**Reply by marcus-j-davies** [reply]
I am using NPM **6.14.11** with nodejs **v14.16.0**

**Reply by lurendrejer** [reply]
same:
root@serv-nodered1:~/.node-red# sudo npm install --production node-red-contrib-zwave-js@latest
npm ERR! code 1
npm ERR! path /root/.node-red/node_modules/node-red-contrib-zwave-js
npm ERR! command failed
npm ERR! command sh -c npm install --build-from-source --production @serialport/bindings@9.2.1 && npm install --build-from-source --production serialport@9.2.1
npm ERR! npm WARN npm npm does not support Node.js v16.10.0
npm ERR! npm WARN npm You should probably upgrade to a newer version of node as we
npm ERR! npm WARN npm can't make any promises that npm will work with this version.
npm ERR! npm WARN npm Supported releases of Node.js are the latest release of 4, 6, 7, 8, 9, 10.
npm ERR! npm WARN npm You can find the latest version at https://nodejs.org/
npm ERR! npm ERR! cb.apply is not a function
npm ERR!
npm ERR! npm ERR! A complete log of this run can be found in:
npm ERR! npm ERR!     /root/.npm/_logs/2021-09-29T19_28_59_789Z-debug.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-09-29T19_29_00_245Z-debug.log

**Reply by marcus-j-davies** [reply]
can you try this?

`sudo npm install --production --ignore-scripts node-red-contrib-zwave-js@latest`

I am not sure how you managed to install the previous version 🤷

**Reply by marcus-j-davies** [reply]
only changes to the install portion was a version bump in **package.json**.
so its strange why you get this.

**Reply by lurendrejer** [reply]
hmm.. i'll look into it tomorrow, I have quite a lot of work to do with my thermostats :)

The last one didn't work either.
I installed via node-red's gui the first time, nothing special..


root@serv-nodered1:~/.node-red# sudo npm install --production --ignore-scripts node-red-contrib-zwave-js@latest

added 2 packages, changed 7 packages, and audited 969 packages in 22s

38 packages are looking for funding
  run `npm fund` for details

34 vulnerabilities (10 low, 10 moderate, 14 high)

To address issues that do not require attention, run:
  npm audit fix

Some issues need review, and may require choosing
a different dependency.

Run `npm audit` for details.
root@serv-nodered1:~/.node-red# systemctl restart nodered

**Reply by marcus-j-davies** [reply]
Ok, I am suspicions about the **seriaports** package.
They seemed to have made a change 

`revert package-locks to npm 6`

It looks as if this is causing problems with npm7 let me know if **--ignore-scripts** allows you to move on.

**Reply by lurendrejer** [reply]
It didn't i ran the command and i'm still running 1.6.0 :)

**Reply by marcus-j-davies** [reply]
oh the  last one worked by the looks of it

**Reply by marcus-j-davies** [reply]
1.6.0? - you mean 6.1.0 😄

**Reply by lurendrejer** [reply]
nopes, sorry :(
![image](https://user-images.githubusercontent.com/30922656/135338008-9be1d2e9-d782-42d9-adad-0f842a9df0c9.png)

**Reply by lurendrejer** [reply]
6.1.0 - i'm slow

**Reply by marcus-j-davies** [reply]
ok what user is nodered running under?

**Reply by lurendrejer** [reply]
root

**Reply by marcus-j-davies** [reply]
ok what does `/root/.node-red/node_modules/node-red-contrib-zwave-js/package.json` show

**Reply by lurendrejer** [reply]
it's fixed - sorry.
Before i ran the --ignore-scripts and restarted nodered afterwards.

I tried stopping node-red, running --ignore-scripts and starting node-red - it shows 6.1.1 now

**Reply by marcus-j-davies** [reply]
yay!

looks like the serialport guys changed something (I'll get onto them about that)

in the mean time you now have a fixed Association Management UI.
Fancy a test (like you did yesterday before shooting off?) - adding a Multi Channel Association

**Reply by lurendrejer** [reply]
yes, will do - right away

**Reply by lurendrejer** [reply]
It accepts the values and the UI starts updating for a binary switch.
I can delete, add, delete, add - and everything happens as it should.

Thank you of the quick fix :)

**Reply by marcus-j-davies** [reply]
We have a node being developed 'cmd-factory' which removes the need to manually write commands 👍

**Comment by lurendrejer** [comment]
This is giving me headaches... 

I've tried sending commands via the gui, but the debugger doesn't show them - maybe because it's a battery device and the get queued.

Is this correct?


let Message = {
    payload: {
        mode: "CCAPI",
        cc: "Thermostat Setpoint",
        method: "set",
        enums: {0:"ThermostatSetpointType"},
        params: ["Heating", 20, celsius]
    }
}
return Message

**Reply by marcus-j-davies** [reply]
> maybe because it's a battery device and the get 

Thats correct - it will get queued, until the device wakes up

**Reply by lurendrejer** [reply]
Well, thats pretty tedious to work with - the binary set-message took me around 30 tries to get right X-)

**Reply by marcus-j-davies** [reply]
in essaance.

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: 2, /* Your node */
        cc: "Binary Switch", /* Your target CC name */
        endpoint: 2, /*Your endpoint if needed */
        method: "set",  /* the CCs Method */
        params: [true] /* params the methods need */
    }
}
return Message
```

here is the CC's
https://zwave-js.github.io/node-zwave-js/#/api/CCs/index

**Reply by marcus-j-davies** [reply]
the new cmd-factory node removes you from that task

**Reply by lurendrejer** [reply]
I got it working:

let Message = {
    payload: {
        mode: "CCAPI",
        cc: "Thermostat Setpoint",
        method: "set",
        enums: { 0: "ThermostatSetpointType" },
        params: ["Heating", 15, "celsius"]
    }
}
return Message


yay!!

**Reply by lurendrejer** [reply]
hmm... according to the link 'scale' should be a number.. doh!

**Reply by marcus-j-davies** [reply]
Nice!

I was just looking.
and scale shows its a number - odd!

https://zwave-js.github.io/node-zwave-js/#/api/CCs/ThermostatSetpoint?id=set

**Reply by marcus-j-davies** [reply]
Snap!

**Reply by marcus-j-davies** [reply]
I dont have a thermo - would you know what scale refers to?

**Reply by marcus-j-davies** [reply]
try 1 😆

**Reply by lurendrejer** [reply]
In the z-wave specs it says 'celsius' - thats where i got it from.
And it is working, the thermostat changed to 15 degrees celsius.

I don't know what to say :)

**Reply by lurendrejer** [reply]
I'll try with Π

**Reply by marcus-j-davies** [reply]
it might be documentation on the zwave-js side in that case (or it wants the numerical value that is sent to the device for celsius)
but does a transform before sending - if that the case, it might just be a documentation issue

good work in sussing it out!

**Reply by lurendrejer** [reply]
It's working and my thermostat-subflow is finished.
Now i can get cracking on the triggers (whos home via unifi, etc.)

thank you up until now - this is really growing on me :)

**Reply by marcus-j-davies** [reply]
nice! (and good to know)
not sure why scale is after a number - but works with a string.

**Reply by marcus-j-davies** [reply]
Let me ask the dev.....

**Reply by marcus-j-davies** [reply]
OK @lurendrejer
"Celsius"  matches into this file

https://github.com/zwave-js/node-zwave-js/blob/6282534a933217dc8241326e24655c5aaf33c07e/packages/config/config/scales.json

so you could in theory use 0, 0x00 (not sure how strict it is) - and it will still use Celsius

**Reply by lurendrejer** [reply]
aah!
i'll stick with celsius, but thank you for investigating :)

**Reply by marcus-j-davies** [reply]
@lurendrejer,

I decided to test this (it was bugging me), and it seems you do indeed need a number.
I think **Celsius** is being used by default - so it looks like it works (and it does but technically using **Celsius** as default)

**using 'celsius'**
```
2021-09-29T21:34:45.980Z DRIVER » [Node 002] [REQ] [SendData]
                                  │ transmit options: 0x25
                                  │ callback id:      6
                                  └─[ThermostatSetpointCCSet]
                                      setpoint type: Heating
                                      value:         15 undefined
```

**using 0x00 (For celsius on that scales lookup)**
```
2021-09-29T21:35:16.757Z DRIVER » [Node 002] [REQ] [SendData]
                                  │ transmit options: 0x25
                                  │ callback id:      7
                                  └─[ThermostatSetpointCCSet]
                                      setpoint type: Heating
                                      value:         15 °C
```

thought you should know - but then if it works anyway (albeit using a default because of a mismatch) - I guess its ok 😄 
but notice the `undefined` in the log

**Reply by lurendrejer** [reply]
Bugger, i'll straighten my course - one day, someone will update a driver and BAM, thankyou - again :)

**Reply by marcus-j-davies** [reply]
Anytime - if not already detected, I like to help new users to the point of getting up and running.

**Comment by marcus-j-davies** [comment]
@lurendrejer,

Can you confirm you currently have a functional setup (I.e no issues) with zwave in node-red?

I am going to switch my focus to the new  `cmd-factory` node (allows to create commands with low code from the user end)
This has been asked by others - so I better do it 😁 

But want to make sure you are at a stage of exploration as opposed to dealing with teething issues?
of course, do holla if you need help with anything.

and I really do suggest joining our discord server for real time discussion, lots of Node-red users, who use the zwave plug-in, apple Homekit plug-in, unify plug-in, deconz and more.

👍

**Reply by lurendrejer** [reply]
Everything is working as expected!
